### PR TITLE
Input group tests and tweaks

### PR DESCRIPTION
**Pull request by mdo**
This PR is adding a new `docs/4.0/examples/tests/` folder to house more complex and dense kitchen sink examples of all the supported variations of a component. Perhaps there's a better location, so I'm open to moving this in the future.

For now, I'm using this to test additional input group tweaks in hopes of fixing our rounded corner issue from #25110. I'll split those commits out after verifying what we support is functioning as intended.

---

screen shot 2018-01-17 at 9 10 11 pm

### CODE REVIEW

**Comment by mdo** [comment]
I've pushed the CSS changes now. Here's where we stand with rendering validation styles in input groups for text inputs and textareas, custom selects, and custom files. Note that custom files cannot have feedback text or tooltips; their markup makes it an immediate no-go.

screen shot 2018-01-17 at 10 20 45 pm

With these complex CSS changes and clever selectors, nothing is lost in promised support so there's technically no backward incompatibility. However, this CSS does clarify just how far we can go with this component given today's CSS selectors.

I was hoping to have this done for v4 stable, but it'll have to roll out after. I'm going to open a PR to make a note of this bug in the docs and link to the tracking issue.

@ysds I know this isn't perfect, but it's the best we can do. These new selectors are the most effective at separately targeting normal inputs/selects/files and validated ones.

**Comment by ysds** [comment]
Let me confirm very soon.

**Comment by ysds** [comment]
The following case seems broken.

```html
<div class="input-group">
  <div class="input-group-prepend">
    <span class="input-group-text">Text</span>
  </div>
  <input type="text" class="form-control is-invalid">
  <div class="input-group-append">
    <span class="input-group-text">Text</span>
  </div>
</div>
<div class="input-group">
  <div class="input-group-prepend"><span class="input-group-text">Text</span></div>
  <select class="custom-select is-invalid">
    <option selected="">Custom select</option>
  </select>
  <div class="input-group-append"><span class="input-group-text">Text</span></div>
</div>
<div class="input-group">
  <div class="input-group-prepend">
    <span class="input-group-text">Text</span>
  </div>
  <input type="text" class="form-control is-valid">
  <div class="valid-feedback">
    Example valid feedback text
  </div>
</div>
<div class="input-group">
  <div class="input-group-prepend"><span class="input-group-text">Text</span></div>
  <select class="custom-select is-valid">
    <option selected="">Custom select</option>
  </select>
  <div class="valid-feedback">
    Example valid feedback text
  </div>
</div>
```

**Comment by mdo** [comment]
Working on those now. I just pushed a commit to resolve the valid feedback one; I had the wrong class in the selector (`.valid` instead of `.is-valid`).

**Comment by ysds** [comment]
The following no feedback-text case are not supported?

```html
<div class="input-group">
  <div class="input-group-prepend">
    <span class="input-group-text">Text</span>
  </div>
  <input type="text" class="form-control is-invalid">
  <div class="input-group-append">
    <span class="input-group-text">Text</span>
  </div>
</div>

<div class="input-group">
  <div class="input-group-prepend">
    <span class="input-group-text">Text</span>
  </div>
  <input type="text" class="form-control is-valid">
  <div class="input-group-append">
    <span class="input-group-text">Text</span>
  </div>
</div>
```

**Comment by ysds** [comment]
Your effort deserves praise. It is up to you to release V4 with some limitations. We have a great utility `rounded-0` :joy:

File: docs/4.0/examples/tests/input-group.md [hunk 1]
```diff
@@ -0,0 +1,462 @@
+---
+layout: default
+title: Test - Input groups
+sizes:
+  - "input-group-sm"
+  - ""
+  - "input-group-lg"
+---
+
+<style>
```
Comment (XhmikosR) [inline docs/4.0/examples/tests/input-group.md]: @mdo: you need to move this to docs CSS file.

File: docs/4.0/examples/tests/input-group.md [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+main { padding: 2rem; }
+.input-group { margin-bottom: .5rem; }
+.row { margin-top: 1rem; margin-bottom: 1.5rem; }
+hr { margin-bottom: 1.5rem; }
+h3, p { margin-bottom: 0; }
+h2 + p { max-width: var(--breakpoint-md); }
+</style>
+
+## Input groups
```
Comment (XhmikosR) [inline docs/4.0/examples/tests/input-group.md]: Please use an empty line after headers.

File: docs/4.0/examples/tests/input-group.md (hunk 1 above)
Comment (mdo) [reply docs/4.0/examples/tests/input-group.md]: Yup yup, just putzing around for now :).

**Comment by mdo** [comment]
Closing as stale for now.

**Pull request closed without merging**